src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+                        }
+
+                        existingCases.Add(builder.MakeTestsForPattern(++index, label.Syntax, rootIdentifier, label.Pattern, label.WhenClause, label.Label));
+                    }
+                }
+            }
+
+            int patternIndex = 0;
+            foreach (BoundSwitchSection section in switchSections)
+            {
+                foreach (BoundSwitchLabel label in section.SwitchLabels)
+                {
+                    if (label.Syntax.Kind() != SyntaxKind.DefaultSwitchLabel)
+                    {
+                        CheckOrAndAndReachability(existingCases, patternIndex, label.Pattern, builder, rootIdentifier, defaultLabel, syntax, diagnostics);
+                        patternIndex++;
+                    }
+                }
+            }
+
+            existingCases.Free();
+        }
+
+        private static void CheckOrAndAndReachability(
+          ArrayBuilder<StateForCase> previousCases,
+          int patternIndex,
+          BoundPattern pattern,
+          DecisionDagBuilder builder,
+          BoundDagTemp rootIdentifier,
+          LabelSymbol defaultLabel,
+          SyntaxNode syntax,
+          BindingDiagnosticBag diagnostics,
+          bool isSwitchExpression = false)
+        {
+            CheckOrReachability(previousCases, patternIndex, pattern,
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+
+            CheckOrReachability(previousCases, patternIndex, MoveNotPatternsDownRewriter.MakeNegatedPattern(pattern),
+                builder, rootIdentifier, defaultLabel, syntax, diagnostics, ignoreDefaultLabel: isSwitchExpression);
+        }
+
+        private static void CheckOrReachability(
+            ArrayBuilder<StateForCase> previousCases,
+            int patternIndex,
+            BoundPattern pattern,
+            DecisionDagBuilder builder,
+            BoundDagTemp rootIdentifier,
+            LabelSymbol defaultLabel,
+            SyntaxNode syntax,
+            BindingDiagnosticBag diagnostics,
+            bool ignoreDefaultLabel = false)
+        {
+            var normalizedPattern = MoveNotPatternsDownRewriter.Rewrite(pattern);
+
+            SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(normalizedPattern);
+            if (setOfOrCases.IsDefault)
+            {
+                return;
+            }
+
+            // We construct a DAG and analyze reachability of branches once per `or` sequence
+            Debug.Assert(!setOfOrCases.IsDefault);
+            foreach (OrCases orCases in setOfOrCases.Set)
+            {
+                using var casesBuilder = TemporaryArray<StateForCase>.GetInstance(orCases.Cases.Count);
+                var labelsToIgnore = PooledHashSet<LabelSymbol>.GetInstance();
+                PopulateStateForCases(builder, rootIdentifier, previousCases, patternIndex, orCases, labelsToIgnore, syntax, ref casesBuilder.AsRef());
```

Comment:
>builder [](http://example.com/codeflow?start=38&length=7)

Is it possible that there is a residual state ins left in the builder from the previous iteration that should be cleared?

Reply:
The state inside the builder is what's passed into the constructor: compilation, default label, `forLowering` flag and discarded diagnostic bag. Those aren't modified, so we're okay.
If you prefer (for safety in case `DecisionDagBuilder` adds more state/mutation), I could create single-use instances. Let me know